nemo_curator/utils/distributed_utils.py

```diff
@@ -617,56 +617,68 @@ def write_to_disk(
 
     Args:
         df: A Dask DataFrame.
-        output_file_dir: The output file path.
+        output_path: The output file path.
         write_to_filename: Boolean representing whether to write the filename using the "filename" column.
         keep_filename_column: Boolean representing whether to keep or drop the "filename" column, if it exists.
         output_type: The type of output file to write. Can be "jsonl" or "parquet".
 
     """
-    if write_to_filename and "filename" not in df.columns:
+
+    # output_path is a file name
+    if isinstance(output_path, str) and output_path.endswith((".jsonl", ".parquet")):
+        _write_to_jsonl_or_parquet(df.compute(), output_path, output_type)
```

Comment:
Couple of things:

1. Directories can also end with `.parquet` so do not think this logic works.
2. Should not call `.compute` and write via client (Too much overhead to bring stuff to client)
3. The only recommended way to write to a single file is repartitioning to a single partition and relying on dask to write 


QQ. Why are we doing this ,  this increases run time by a factor of num_workers ?

Reply:
@VibhuJawa it [came from here](https://docs.google.com/document/d/19lnrlC7k9lR0PM_xB6M7JXQCa-qRG97USmUPwqe_4u4/edit?tab=t.0#heading=h.dwj89051ed5v)

Though now that I read the requirements more clearly, I don't know if we should support it too, since any distributed data processing system like dask (or spark), will write using multiple workers, hence writing to the same file will always be inefficient since you'll have to perform `.compute() / .collect()` (or `repartition(1)`) to get the data to the one node which will perform the write. 

In my previous job we also had a need for something similar, where we made it a) hard b) very explicit.

An example could be 

```python
DocumentDataset.to_json(..) # write distributed fashion
DocumentDataset.to_single_file(filetype=...) # OOM is possible
```

Let me know what you think?


Also for @VibhuJawa, assuming two things
1. client specs is same as any workers spec
2. the operation after repartition(1) is write (no intermediate map_partition)

Then isn't `.repartition(1)` and `.compute(..)` bottlenecked by the worker (or client)'s memory since the data will need to fit in that one single node, i.e has potential to OOM?

Reply:
Yeah, I agree using `compute()` is bad, but I also think it makes sense to allow users to write to a single file if they really want to do it. I imagine this would only happen if the dataset they are working with is pretty small anyway.

I like Praateek's idea of calling it a separate function like `to_single_file` or perhaps adding a parameter to the `to_json` and `to_parquet` function which the user has to explicitly set to write a single file.

Reply:
> client specs is same as any workers spec

Yes. 

> I like Praateek's idea of calling it a separate function like to_single_file or perhaps adding a parameter to the to_json and to_parquet function which the user has to explicitly set to write a single file.

Yeah, we can have `to_single_file` , i think that is still tenable, we should have a warning though if we are doing a repartiton from npartitons > 1 to a single partition.

Reply:
Looking through the original discussion where this came up I see the comment: `with small datasets with a single partition, this save API is difficult to use`.

Maybe for @Maghoumi use case, just enabling the `to_single_file` method by default if ddf is a single partition (no matter how big or small) is good enough? Is that the case?

Reply:
Sarah and I had a quick discussion offline. She told me the new way this is done is such that if the dataframe is a single partition, then it writes to a single file. If it's not a single partition, then a runtime warning is issued, the output path is treated as a directory and the data set will be written in multiple files (one for each partition).

Having read this discussion and the one below, I think this is the most reasonable way of handling it and won't require adding another API such as `to_single_file`.